### PR TITLE
Never ignore JitFramed flag

### PR DESCRIPTION
**Pull request by EgorBo**
Contributes to https://github.com/dotnet/runtime/issues/105690

### CODE REVIEW

**Comment by EgorBo** [comment]
Run 0:
@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;

public class MyBench
{
    static MyObj[] Data = TestData();
    static MemoryStream Output = new(new byte[1024 * 1024]);
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    [ArgumentsSource(nameof(TestData))]
    public void JsonStatham(MyObj[] arrays) => JsonSerializer.Serialize(Output, Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
Run 1:
@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;

public class MyBench
{
    static MyObj[] Data = TestData();
    static MemoryStream Output = new(new byte[1024 * 1024]);
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    [ArgumentsSource(nameof(TestData))]
    public void JsonStatham(MyObj[] arrays) => JsonSerializer.Serialize(Output, Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
Run 0:
@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
Run 1:
@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by jkotas** [comment]
Do you understand why this produces better stacktraces?

**Comment by EgorBo** [comment]
> Do you understand why this produces better stacktraces?

It just seemed odd to me that JIT just ignores `DOTNET_JitFramed=1` (so code is optimized, but literally nothing is inlined in both base and diff). I already can see a difference e.g. in `Grisu::TryRonShortest`

image

it reproduces for both Intel and Amd and for both runs.


Related: https://www.brendangregg.com/blog/2024-03-17/the-return-of-the-frame-pointers.html

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -4151,14 +4151,11 @@ void Compiler::compSetOptimizationLevel()
             codeGen->setFrameRequired(true);
         }
 
-#if !defined(TARGET_AMD64)
         // The VM sets JitFlags::JIT_FLAG_FRAMED for two reasons: (1) the DOTNET_JitFramed variable is set, or
         // (2) the function is marked "noinline". The reason for #2 is that people mark functions
-        // noinline to ensure the show up on in a stack walk. But for AMD64, we don't need a frame
```
Comment (jkotas) [inline src/coreclr/jit/compiler.cpp:4157]: The deleted comment is still true on Windows. I guess it was written before we cared about non-Windows.

**Comment by jkotas** [comment]
> It just seemed odd to me that JIT just ignores DOTNET_JitFramed=1

It makes on Windows x64. The RBP-frames are useless on Windows x64.

I would like to understand why this helps. Our strategy for omitting the RBP-frames should not mess with RBP-based stackwalking on Windows x86 or Linux x64: We do not use EBP/RBP as a general purpose register, so the effect of methods without the frame should be similar to inlining or tailcalling. It should not break the stackwalking.

**Comment by EgorBo** [comment]
Run 0:
@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
Run with JitNoInline=0

@EgorBot -amd -intel -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
Here are the runs with completely default settings (no attempts to stop any inlining)

Intel (Main):
- Run 0: https://telegafiles.blob.core.windows.net/telega/base_flamegraph_7f8d8cc0.svg
- Run 1: https://telegafiles.blob.core.windows.net/telega/base_flamegraph_638892f8.svg

Intel (**this PR**):
- Run 0: https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_7f8d8cc0.svg
- Run 1: https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_638892f8.svg

Amd (Main):
- Run 0: https://telegafiles.blob.core.windows.net/telega/base_flamegraph_c6d6e9ed.svg
- Run 1: https://telegafiles.blob.core.windows.net/telega/base_flamegraph_8e363d9e.svg

Amd (**this PR**):
- Run 0: https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_c6d6e9ed.svg
- Run 1: https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_8e363d9e.svg

I think I cannot confirm that my change "fix" the flamegraphs, e.g. AMD (this PR) two still look random (while Main is ok). I guess I need to find the problem elsewhere, e.g. collect traces and compare inlining/tailcall decisions etc.

**Comment by EgorBo** [comment]
@EgorBot -arm64 -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
@EgorBot -arm64 -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
4 flamegraphs for arm64:

- https://telegafiles.blob.core.windows.net/telega/base_flamegraph_d0f91da0.svg
- https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_d0f91da0.svg
- https://telegafiles.blob.core.windows.net/telega/base_flamegraph_f5f64bd5.svg
- https://telegafiles.blob.core.windows.net/telega/diff_flamegraph_f5f64bd5.svg

the last one has differences 🤔

**Comment by EgorBo** [comment]
I wonder if it's possible to add Tier name to symbols..

**Comment by EgorBo** [comment]
Test run with `DOTNET_PerfMapShowOptimizationTiers=1`

@EgorBot -arm64 -profiler
```cs
using System;
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<MyBench>(args: args);

public class MyBench
{
    static MyObj[] Data = TestData();
    static TimeSpan TS = TimeSpan.FromDays(42);

    public static MyObj[] TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), 3.14, TS, null);
            MyObj obj2 = new("'')((*&&^%$@#$%$^&*())''';(*&^%$E##^%$&%^*(", i, Guid.NewGuid(), 3.14, TS, obj1);
            testData[i] = obj2;
        }
        return testData;
    }

    [Benchmark]
    public object JsonStatham() => JsonSerializer.Serialize(Data);
}

public record MyObj(string Name, int Age, Guid Id, double SomeFloat, TimeSpan Ts, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
@EgorBot -amd
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Comment by EgorBo** [comment]
@EgorBot -intel -keep
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd -keep -profiler
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Comment by EgorBo** [comment]
@EgorBot -amd -intel -profiler
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Comment by EgorBo** [comment]
@EgorBot -arm64 -profiler
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Comment by EgorBo** [comment]
@EgorBot -intel -amd -profiler
```cs
using System;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class Bench
{
    [Benchmark]
    public void WB()
    {
        Foo foo = new Foo();
        for (long i = 0; i < 200000000; i++)
            foo.x = foo;
    }
}

internal class Foo
{
    public volatile Foo x;
}
```

**Pull request closed without merging**